sdk/src/Services/S3/Custom/Transfer/Internal/BufferedPartDataHandler.cs

```diff
@@ -0,0 +1,165 @@
+        }
+        
+        public void ReleaseCapacity()
+        {
+            _partBufferManager.ReleaseBufferSpace();
+        }
+        
+        public void OnDownloadComplete(Exception exception)
+        {
+            _partBufferManager.MarkDownloadComplete(exception);
+        }
+        
+        public void Dispose()
+        {
+            // _partBufferManager is owned by caller, don't dispose
+        }
+        
+        private async Task<StreamPartBuffer> BufferPartFromResponseAsync(
+            int partNumber, 
+            GetObjectResponse response, 
+            CancellationToken cancellationToken)
+        {
+            StreamPartBuffer downloadedPart = null;
+            byte[] partBuffer = null;
+            
+            try
+            {
+                // Use ContentLength to determine exact bytes to read and allocate
+                long expectedBytes = response.ContentLength;
+                int initialBufferSize = (int)expectedBytes;
+                
+                partBuffer = ArrayPool<byte>.Shared.Rent(initialBufferSize);
+                
+                int totalRead = 0;
+                
+                // Read response stream into buffer based on ContentLength
+                while (totalRead < expectedBytes)
+                {
+                    // Calculate how many bytes we still need to read
+                    int remainingBytes = (int)(expectedBytes - totalRead);
+                    int bufferSpace = partBuffer.Length - totalRead;
+                    
+                    // Expand buffer if needed BEFORE reading
+                    if (bufferSpace == 0)
+                    {
+                        partBuffer = ExpandPartBuffer(partBuffer, totalRead);
+                        bufferSpace = partBuffer.Length - totalRead;
+                    }
+                    
+                    // Read size is minimum of: remaining bytes needed, buffer space available, and configured buffer size
+                    int readSize = Math.Min(Math.Min(remainingBytes, bufferSpace), _config.BufferSize);
+                    
+                    // Read directly into final destination at correct offset
+                    int bytesRead = await response.ResponseStream.ReadAsync(
+                        partBuffer,     // Destination: final ArrayPool buffer
+                        totalRead,      // Offset: current position in buffer
+                        readSize,       // Count: optimal chunk size
+                        cancellationToken).ConfigureAwait(false);
+                    
+                    if (bytesRead == 0) 
+                    {
+                        // Unexpected EOF
+                        break;
+                    }
+                    
+                    totalRead += bytesRead;
+                }
+
+                // Create ArrayPool-based StreamPartBuffer (no copying!)
+                downloadedPart = StreamPartBuffer.FromArrayPoolBuffer(
```

Comment:
I think there is some logic issues with the `partBuffer` renting and returning because the `partBuffer` is exposed on the `downloadPart` via the  `ArrayPoolBuffer`. The `downloadPart` instances is returned to be used else where in the flow reading from `ArrayPoolBuffer` but the underlying `partBuffer` byte array is being returned to the pool in the `finally` block.